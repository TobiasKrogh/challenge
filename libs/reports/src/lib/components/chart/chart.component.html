<ng-container *ngIf="data$ | async as data">
  <div class="legend">
    <span
      *ngFor="let entry of data"
      [ngStyle]="{
        '--color': entry.color
      }"
      >{{ entry.label }}</span
    >
  </div>

  <div
    class="pie"
    [ngStyle]="{
      'background-image': pie$ | async
    }"
  >
    <span
      *ngFor="let entry of data"
      [ngStyle]="{
        transform: entry.labelTransform
      }"
    >
      {{ entry.percentage }}%
    </span>
  </div>
</ng-container>
