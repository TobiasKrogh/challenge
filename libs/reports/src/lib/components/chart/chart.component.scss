:host {
  display: block;
}

.legend {
  display: flex;
  justify-content: space-around;

  span {
    font: var(--font-chart-legend);
    padding-left: 1.75rem;
    position: relative;

    &::before {
      background-color: var(--color);
      border-radius: var(--border-radius--default);
      bottom: 0;
      content: '';
      left: 0;
      position: absolute;
      top: 0;
      width: 1rem;
    }
  }
}

.pie {
  border-radius: 50%;
  height: 16.875rem;
  margin: 2rem auto;
  position: relative;
  width: 16.875rem;

  &::after {
    background-color: var(--hex-color-white);
    border-radius: 50%;
    content: '';
    height: 8.75rem;
    left: 4.0625rem;
    position: absolute;
    top: 4.0625rem;
    width: 8.75rem;
  }

  span {
    color: var(--hex-color-white);
    display: block;
    font: var(--font-chart-label);
    height: 1.25rem;

    position: absolute;
    top: 50%;
    left: 50%;
    margin-left: -1.25rem;
    margin-top: -0.625rem;
    text-align: center;
    width: 2.5rem;
  }
}
