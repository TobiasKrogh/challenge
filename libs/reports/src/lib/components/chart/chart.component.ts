import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import {
  combineLatest,
  map,
  Observable,
  ReplaySubject,
  shareReplay,
} from 'rxjs';
import { IChartData } from '../../model/chart-slice.interface';
import { IReportByGateway } from '../../model/report-by-gateway.interface';
import { IReportByProject } from '../../model/report-by-project.interface';

type TChartSource = IReportByGateway | IReportByProject;

const COLORS = ['#A259FF', '#6497B1', 'FFC107', '#F24E1E', '#1BC5BD'];

@Component({
  selector: 'reports-chart',
  templateUrl: './chart.component.html',
  styleUrls: ['./chart.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ChartComponent {
  private sources$ = new ReplaySubject<TChartSource[]>(1);
  private total$ = new ReplaySubject<number>(1);

  @Input() set sources(data: TChartSource[] | undefined) {
    if (data) {
      this.sources$.next(data);
    }
  }

  @Input() set total(amount: number) {
    if (amount) {
      this.total$.next(amount);
    }
  }

  protected readonly data$: Observable<IChartData[]> = combineLatest([
    this.sources$,
    this.total$,
  ]).pipe(
    map(([sources, total]) => {
      let lastPercentage = 0;
      return sources.map((source, index) => {
        let label = '';

        if (this.isReportByGateway(source)) {
          label = source.gateway.name;
        }

        if (this.isReportByProject(source)) {
          label = source.project.name;
        }

        const percentage = Math.round((source.total / total) * 100);

        const degree = Math.max(
          0,
          Math.round(this.getDegree(lastPercentage + percentage)) - 180
        );
        lastPercentage = percentage;

        let adjustedDegree = Math.max(0, degree - 10);
        if (degree >= 180) {
          adjustedDegree = degree + 180 - this.getDegree(percentage) - 10;
        }

        return {
          color: COLORS[index],
          degree,
          label,
          labelTransform: `rotate(${adjustedDegree}deg) translate(104px) rotate(-${adjustedDegree}deg)`,
          percentage,
        };
      });
    }),
    shareReplay(1)
  );

  protected readonly pie$ = this.data$.pipe(
    map((data) => {
      let current = '';
      const slices = data
        .map(({ color, percentage }) => {
          const degree = this.getDegree(percentage);
          const slice = `${color} ${current}${degree}deg`;
          current = `${degree}deg `;
          return slice;
        })
        .join(', ');
      return `conic-gradient(${slices})`;
    }),
    shareReplay(1)
  );

  private isReportByGateway(source: TChartSource): source is IReportByGateway {
    return 'gateway' in source;
  }

  private isReportByProject(source: TChartSource): source is IReportByProject {
    return 'project' in source;
  }

  private getDegree(percentage: number): number {
    return (360 / 100) * percentage;
  }
}
