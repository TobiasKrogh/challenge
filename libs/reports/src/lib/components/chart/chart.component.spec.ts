import { ComponentFixture, TestBed } from '@angular/core/testing';
import { cold } from 'jest-marbles';
import {
  MOCK_GATEWAY,
  MOCK_PROJECT,
  MOCK_REPORT,
} from '../../../testing/data.mocks';

import { ChartComponent } from './chart.component';

describe('ChartComponent', () => {
  let component: ChartComponent;
  let fixture: ComponentFixture<ChartComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ChartComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(ChartComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  describe('When rendering', () => {
    it('Then it matches snapshot (no data provided)', () => {
      fixture.detectChanges();
      expect(fixture.debugElement).toMatchSnapshot();
    });

    describe('And valid data was provided (gateways)', () => {
      beforeEach(() => {
        component.total = 50;
        component.sources = [
          {
            data: [MOCK_REPORT],
            gateway: MOCK_GATEWAY,
            open: true,
            total: 27,
          },
          {
            data: [MOCK_REPORT],
            gateway: MOCK_GATEWAY,
            open: true,
            total: 23,
          },
        ];
      });

      it('Then it matches snapshot (data provided - projects)', () => {
        fixture.detectChanges();
        expect(fixture.debugElement).toMatchSnapshot();
      });

      it('Then gradient is properly set up', () => {
        expect(component['pie$']).toBeObservable(
          cold('a', {
            a: 'conic-gradient(#A259FF 194.4deg, #6497B1 194.4deg 165.6deg)',
          })
        );
      });
    });

    describe('And valid data was provided (projects)', () => {
      beforeEach(() => {
        component.total = 50;
        component.sources = [
          {
            data: [MOCK_REPORT],
            project: MOCK_PROJECT,
            open: true,
            total: 27,
          },
          {
            data: [MOCK_REPORT],
            project: MOCK_PROJECT,
            open: true,
            total: 23,
          },
        ];
      });

      it('Then it matches snapshot (data provided - projects)', () => {
        fixture.detectChanges();
        expect(fixture.debugElement).toMatchSnapshot();
      });

      it('Then gradient is properly set up', () => {
        expect(component['pie$']).toBeObservable(
          cold('a', {
            a: 'conic-gradient(#A259FF 194.4deg, #6497B1 194.4deg 165.6deg)',
          })
        );
      });
    });
  });
});
