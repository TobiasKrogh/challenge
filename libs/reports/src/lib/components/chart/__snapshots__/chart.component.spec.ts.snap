// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartComponent When rendering And valid data was provided (gateways) Then it matches snapshot (data provided - projects) 1`] = `
DebugElement {
  "nativeNode": <div
    id="root2"
  >
    <div
      class="legend"
    >
      <span
        style="--color: #A259FF;"
      >
        name
      </span>
      <span
        style="--color: #6497B1;"
      >
        name
      </span>
      
    </div>
    <div
      class="pie"
    >
      <span
        style="transform: rotate(4deg) translate(104px) rotate(-4deg);"
      >
         54% 
      </span>
      <span
        style="transform: rotate(184.4deg) translate(104px) rotate(-184.4deg);"
      >
         46% 
      </span>
      
    </div>
    
    
  </div>,
}
`;

exports[`ChartComponent When rendering And valid data was provided (projects) Then it matches snapshot (data provided - projects) 1`] = `
DebugElement {
  "nativeNode": <div
    id="root4"
  >
    <div
      class="legend"
    >
      <span
        style="--color: #A259FF;"
      >
        name
      </span>
      <span
        style="--color: #6497B1;"
      >
        name
      </span>
      
    </div>
    <div
      class="pie"
    >
      <span
        style="transform: rotate(4deg) translate(104px) rotate(-4deg);"
      >
         54% 
      </span>
      <span
        style="transform: rotate(184.4deg) translate(104px) rotate(-184.4deg);"
      >
         46% 
      </span>
      
    </div>
    
    
  </div>,
}
`;

exports[`ChartComponent When rendering Then it matches snapshot (no data provided) 1`] = `
DebugElement {
  "nativeNode": <div
    id="root1"
  >
    
  </div>,
}
`;
